flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/StopWithSavepointTerminationHandlerImpl.java

```diff
@@ -0,0 +1,258 @@
+
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+
+import javax.annotation.Nonnull;
+
+import java.util.Collection;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * {@code StopWithSavepointTerminationHandlerImpl} implements {@link
+ * StopWithSavepointTerminationHandler}.
+ *
+ * <p>The operation only succeeds if both steps, the savepoint creation and the successful
+ * termination of the job, succeed. If the former step fails, the operation fails exceptionally
+ * without any further actions. If the latter one fails, a global fail-over is triggered before
+ * failing the operation.
+ */
+public class StopWithSavepointTerminationHandlerImpl
+        implements StopWithSavepointTerminationHandler {
+
+    private final Logger log;
+
+    private final SchedulerNG scheduler;
+    private final CheckpointScheduling checkpointScheduling;
+    private final JobID jobId;
+
+    private final CompletableFuture<String> result = new CompletableFuture<>();
+
+    private State state = new WaitingForSavepoint();
+
+    public <S extends SchedulerNG & CheckpointScheduling> StopWithSavepointTerminationHandlerImpl(
+            @Nonnull JobID jobId, @Nonnull S schedulerWithCheckpointing, @Nonnull Logger log) {
+        this(jobId, schedulerWithCheckpointing, schedulerWithCheckpointing, log);
+    }
+
+    @VisibleForTesting
+    StopWithSavepointTerminationHandlerImpl(
+            @Nonnull JobID jobId,
+            @Nonnull SchedulerNG scheduler,
+            @Nonnull CheckpointScheduling checkpointScheduling,
+            @Nonnull Logger log) {
+        this.jobId = checkNotNull(jobId);
+        this.scheduler = checkNotNull(scheduler);
+        this.checkpointScheduling = checkNotNull(checkpointScheduling);
+        this.log = checkNotNull(log);
+    }
+
+    @Override
+    public CompletableFuture<String> handlesStopWithSavepointTermination(
+            CompletableFuture<CompletedCheckpoint> completedSavepointFuture,
+            CompletableFuture<Collection<ExecutionState>> terminatedExecutionsFuture,
+            ComponentMainThreadExecutor mainThreadExecutor) {
+        completedSavepointFuture
+                .whenCompleteAsync(
+                        (completedSavepoint, throwable) -> {
+                            if (throwable != null) {
+                                handleSavepointCreationFailure(throwable);
+                            } else {
+                                handleSavepointCreation(completedSavepoint);
+                            }
+                        },
+                        mainThreadExecutor)
+                .thenCompose(
+                        aVoid ->
+                                terminatedExecutionsFuture.thenAcceptAsync(
+                                        this::handleExecutionsTermination, mainThreadExecutor));
```

Comment:
I think this logic does not really belong to this class. If the class states that one first needs to signal the savepoint status and then only the termination status, then it should be good enough. Moreover, by not doing this part in this class, the class does not need to know about the concept of a main thread.

Reply:
Good point. I refactored it and introduced a `StopWithSavepointTerminationManager` to enforce the right execution order.